src/Microsoft.ApplicationInsights/PreciseTimestamp.cs

```diff
@@ -0,0 +1,109 @@
+﻿namespace Microsoft.ApplicationInsights
+{
+    using System;
+    using System.Diagnostics.CodeAnalysis;
+#if NET45 || NET46
+    using System.Diagnostics;
+    using System.Threading;
+#endif
+
+    internal class PreciseTimestamp
+    {
+        private static readonly object Lck = new object();
+        private static PreciseTimestamp instance = null;
+
+#if NET45 || NET46
+        private static TimeSync timeSync;
+        private readonly Timer syncTimeUpdater;
+
+        private PreciseTimestamp()
+        {
+            this.syncTimeUpdater = InitializeSyncTimer();
+        }
+#endif
+
+        public static PreciseTimestamp Instance
+        {
+            get
+            {
+                if (instance == null)
+                {
+                    lock (Lck)
+                    {
+                        if (instance == null)
+                        {
+                            instance = new PreciseTimestamp();
+                        }
+                    }
+                }
+
+                return instance;
+            }
+        }
+
+        /// <summary>
+        /// Returns high resolution (1 DateTime tick) current UTC DateTime. 
+        /// </summary>
+        [SuppressMessage("Microsoft.Performance", "CA1822:MarkMembersAsStatic", Justification = "Enforcing instance fields initialization.")]
+        internal DateTimeOffset GetUtcNow()
+        {
+#if NET45 || NET46
+            // DateTime.UtcNow accuracy on .NET Framework is ~16ms, this method 
+            // uses combination of Stopwatch and DateTime to calculate accurate UtcNow.
+
+            var tmp = timeSync;
+
+            // Timer ticks need to be converted to DateTime ticks
+            long dateTimeTicksDiff = (long)((Stopwatch.GetTimestamp() - tmp.SyncStopwatchTicks) * 10000000L /
+                                            (double)Stopwatch.Frequency);
```

Comment:
This expression is captured in `StopwatchTicksToTimeSpanTicks` in `OperationTelemetryExtensions`. Can we share the two implementations?